File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n=s.length();
        int left=0;
        int ans=0;
        HashMap<Character,Integer> map=new HashMap<>();
        for(int i=0;i<n;i++)
        {
            if(!map.containsKey(s.charAt(i)))
            {
                map.put(s.charAt(i),i);
                ans=Math.max(ans,i-left+1);
            }
            else
            {
                left=Math.max(left,map.get(s.charAt(i))+1);
                ans=Math.max(ans,i-left+1);
                map.put(s.charAt(i),i);
            }
        }
        return ans;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n=len(s)
        left=0
        ans=0
        map={}
        for i in range(n):
            if s[i] not in map:
                map[s[i]]=i
                ans=max(ans,i-left+1)
            else:
                left=max(left,map[s[i]]+1)
                ans=max(ans,i-left+1)
                map[s[i]]=i
        return ans